src/useQueuedForceUpdate.ts

```diff
@@ -0,0 +1,41 @@
+import React from "react"
+
+export type ForceUpdate = () => void
+
+let insideRender: boolean = false
+let forceUpdateQueue: ForceUpdate[] = []
+
+export function useQueuedForceUpdate(forceUpdate: ForceUpdate): ForceUpdate {
+    return () => {
+        if (insideRender) {
+            forceUpdateQueue.push(forceUpdate)
+        } else {
+            forceUpdate()
+        }
+    }
+}
+
+export function useQueuedForceUpdateBlock<T>(callback: () => T) {
+    // start intercepting force-update calls
+    insideRender = true
+    forceUpdateQueue = []
+    try {
+        const result = callback()
+
+        // stop intercepting force-update
+        insideRender = false
+        // store queue or nothing if it was empty to execute useLayoutEffect only when necessary
+        const queue = forceUpdateQueue.length > 0 ? forceUpdateQueue : undefined
```

Comment:
Don't we need a copy here?

Reply:
The copy is stored in the `queue` const and `forceUpdateQueue` is always initialized to an empty array before the next render so it should be fine.